File: desktop-app-vue/src/main/ai-engine/smart-plan-cache.js
/**
 * SmartPlanCache - 智能任务计划缓存
 *
 * 使用语义相似度匹配历史任务计划，大幅提升缓存命中率
 *
 * 核心功能:
 * 1. LLM Embedding向量化
 * 2. 余弦相似度匹配
 * 3. LRU淘汰策略
 * 4. 统计和监控
 *
 * @module ai-engine/smart-plan-cache
 */

const { logger } = require('../utils/logger.js');
const { v4: uuidv4 } = require('uuid');

/**
 * LRU缓存条目
 */
class CacheEntry {
  constructor(key, request, plan, embedding) {
    this.id = uuidv4().slice(0, 8);
    this.key = key; // 原始请求的hash（快速查找）
    this.request = request; // 原始请求文本
    this.plan = plan; // 任务计划
    this.embedding = embedding; // 请求的embedding向量
    this.hits = 0; // 命中次数
    this.createdAt = Date.now();
    this.lastHitAt = null;
    this.lastAccessAt = Date.now();
  }
}

/**
 * SmartPlanCache 类
 */
class SmartPlanCache {
  constructor(options = {}) {
    this.maxSize = options.maxSize || 1000; // 最大缓存条目数
    this.similarityThreshold = options.similarityThreshold || 0.85; // 相似度阈值（0-1）
    this.ttl = options.ttl || 7 * 24 * 60 * 60 * 1000; // 7天TTL
    this.enabled = options.enabled !== false;
    this.llmManager = options.llmManager; // LLM管理器（用于embedding）

    // 缓存存储: Map<key, CacheEntry>
    this.cache = new Map();

    // 访问顺序（LRU）: Array<key>
    this.accessOrder = [];

    // 统计信息
    this.stats = {
      totalRequests: 0,
      cacheHits: 0,
      cacheMisses: 0,
      semanticHits: 0, // 语义匹配命中
      exactHits: 0, // 精确匹配命中
      evictions: 0, // 淘汰次数
      embeddingCalls: 0, // embedding调用次数
      embeddingFailures: 0,
    };

    // 定期清理过期条目
    this.cleanupInterval = setInterval(() => {
      this._cleanupExpired();
    }, 60 * 60 * 1000); // 每小时清理一次

    logger.info('[SmartPlanCache] 智能计划缓存已初始化', {
      maxSize: this.maxSize,
      similarityThreshold: this.similarityThreshold,
      ttl: `${this.ttl / 1000 / 60 / 60 / 24}天`,
    });
  }

  /**
   * 获取缓存（支持语义匹配）
   * @param {string} request - 用户请求
   * @returns {Promise<Object|null>} 缓存的计划或null
   */
  async get(request) {
    if (!this.enabled) {
      return null;
    }

    this.stats.totalRequests++;

    // 1. 尝试精确匹配（快速路径）
    const exactKey = this._hash(request);
    const exactEntry = this.cache.get(exactKey);

    if (exactEntry && !this._isExpired(exactEntry)) {
      this._recordHit(exactEntry, 'exact');
      this.stats.exactHits++;
      this.stats.cacheHits++;

      logger.debug(`[SmartPlanCache] ✅ 精确命中: ${exactEntry.id}, 命中次数: ${exactEntry.hits}`);

      return exactEntry.plan;
    }

    // 2. 语义相似度匹配
    const semanticPlan = await this._semanticSearch(request);

    if (semanticPlan) {
      this.stats.semanticHits++;
      this.stats.cacheHits++;
      return semanticPlan;
    }

    // 3. 缓存未命中
    this.stats.cacheMisses++;
    logger.debug('[SmartPlanCache] ❌ 缓存未命中');

    return null;
  }

  /**
   * 设置缓存
   * @param {string} request - 用户请求
   * @param {Object} plan - 任务计划
   */
  async set(request, plan) {
    if (!this.enabled) {
      return;
    }

    const key = this._hash(request);

    // 检查是否已存在
    if (this.cache.has(key)) {
      logger.debug(`[SmartPlanCache] 更新现有缓存: ${key}`);
      const entry = this.cache.get(key);
      entry.plan = plan;
      entry.lastAccessAt = Date.now();
      this._updateAccessOrder(key);
      return;
    }

    // 计算embedding
    const embedding = await this._getEmbedding(request);

    if (!embedding) {
      logger.warn('[SmartPlanCache] Embedding失败，跳过缓存');
      return;
    }

    // 创建新条目
    const entry = new CacheEntry(key, request, plan, embedding);

    // 检查缓存是否已满
    if (this.cache.size >= this.maxSize) {
      this._evictLRU();
    }

    // 添加到缓存
    this.cache.set(key, entry);
    this.accessOrder.push(key);

    logger.debug(`[SmartPlanCache] ✅ 缓存已添加: ${entry.id}, 缓存大小: ${this.cache.size}/${this.maxSize}`);
  }

  /**
   * 语义搜索
   * @private
   */
  async _semanticSearch(request) {
    // 计算请求的embedding
    const requestEmbedding = await this._getEmbedding(request);

    if (!requestEmbedding) {
      return null;
    }

    let bestMatch = null;
    let bestSimilarity = 0;

    // 遍历所有缓存条目，计算相似度
    for (const entry of this.cache.values()) {
      if (this._isExpired(entry)) {
        continue;
      }

      const similarity = this._cosineSimilarity(requestEmbedding, entry.embedding);

      if (similarity > bestSimilarity && similarity >= this.similarityThreshold) {
        bestSimilarity = similarity;
        bestMatch = entry;
      }
    }

    if (bestMatch) {
      this._recordHit(bestMatch, 'semantic');
      logger.debug(`[SmartPlanCache] ✅ 语义命中: ${bestMatch.id}, 相似度: ${(bestSimilarity * 100).toFixed(2)}%, 命中次数: ${bestMatch.hits}`);
      return bestMatch.plan;
    }

    return null;
  }

  /**
   * 获取文本的embedding向量
   * @private
   */
  async _getEmbedding(text) {
    if (!this.llmManager) {
      logger.warn('[SmartPlanCache] LLM管理器未配置，无法生成embedding');
      return null;
    }

    this.stats.embeddingCalls++;

    try {
      // 尝试调用LLM的embedding API
      // 注意：这需要LLM管理器支持embedding功能
      if (typeof this.llmManager.getEmbedding === 'function') {
        const embedding = await this.llmManager.getEmbedding(text);
        return embedding;
      }

      // 如果LLM管理器不支持embedding，使用简单的TF-IDF向量化作为后备
      logger.warn('[SmartPlanCache] LLM管理器不支持embedding，使用TF-IDF后备方案');
      return this._simpleTFIDFVector(text);

    } catch (error) {
      this.stats.embeddingFailures++;
      logger.error('[SmartPlanCache] Embedding生成失败:', error.message);
      return null;
    }
  }

  /**
   * 简单的TF-IDF向量化（后备方案）
   * @private
   */
  _simpleTFIDFVector(text) {
    // 分词（简单空格分割）
    const tokens = text.toLowerCase().split(/\s+/).filter(t => t.length > 2);

    // 计算词频
    const freq = {};
    for (const token of tokens) {
      freq[token] = (freq[token] || 0) + 1;
    }

    // 归一化（简化版）
    const totalTokens = tokens.length;
    const vector = [];

    // 使用固定词汇表（简化实现）
    const vocab = this._getVocab();

    for (const word of vocab) {
      const tf = (freq[word] || 0) / totalTokens;
      vector.push(tf);
    }

    return vector;
  }

  /**
   * 获取词汇表（简化实现）
   * @private
   */
  _getVocab() {
    // 简化的词汇表（实际应该是从所有请求中提取）
    return [
      'create', 'add', 'build', 'generate', 'implement',
      'update', 'modify', 'change', 'edit', 'refactor',
      'delete', 'remove', 'clean', 'clear',
      'test', 'check', 'validate', 'verify',
      'file', 'code', 'function', 'class', 'component',
      'api', 'database', 'server', 'client',
      'feature', 'bug', 'fix', 'improve', 'optimize',
    ];
  }

  /**
   * 余弦相似度计算
   * @private
   */
  _cosineSimilarity(vec1, vec2) {
    if (!vec1 || !vec2 || vec1.length !== vec2.length) {
      return 0;
    }

    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;

    for (let i = 0; i < vec1.length; i++) {
      dotProduct += vec1[i] * vec2[i];
      norm1 += vec1[i] * vec1[i];
      norm2 += vec2[i] * vec2[i];
    }

    const denominator = Math.sqrt(norm1) * Math.sqrt(norm2);

    if (denominator === 0) {
      return 0;
    }

    return dotProduct / denominator;
  }

  /**
   * 记录缓存命中
   * @private
   */
  _recordHit(entry, type = 'unknown') {
    entry.hits++;
    entry.lastHitAt = Date.now();
    entry.lastAccessAt = Date.now();
    this._updateAccessOrder(entry.key);
  }

  /**
   * 更新访问顺序（LRU）
   * @private
   */
  _updateAccessOrder(key) {
    // 从旧位置移除
    const index = this.accessOrder.indexOf(key);
    if (index !== -1) {
      this.accessOrder.splice(index, 1);
    }

    // 添加到末尾（最近访问）
    this.accessOrder.push(key);
  }

  /**
   * LRU淘汰
   * @private
   */
  _evictLRU() {
    if (this.accessOrder.length === 0) {
      return;
    }

    // 淘汰最久未使用的条目
    const lruKey = this.accessOrder.shift();
    const evictedEntry = this.cache.get(lruKey);

    if (evictedEntry) {
      this.cache.delete(lruKey);
      this.stats.evictions++;

      logger.debug(`[SmartPlanCache] 🗑️ LRU淘汰: ${evictedEntry.id}, 命中次数: ${evictedEntry.hits}`);
    }
  }

  /**
   * 检查条目是否过期
   * @private
   */
  _isExpired(entry) {
    return Date.now() - entry.createdAt > this.ttl;
  }

  /**
   * 清理过期条目
   * @private
   */
  _cleanupExpired() {
    const now = Date.now();
    let cleanedCount = 0;

    for (const [key, entry] of this.cache.entries()) {
      if (this._isExpired(entry)) {
        this.cache.delete(key);
        const index = this.accessOrder.indexOf(key);
        if (index !== -1) {
          this.accessOrder.splice(index, 1);
        }
        cleanedCount++;
      }
    }

    if (cleanedCount > 0) {
      logger.info(`[SmartPlanCache] 🧹 清理过期条目: ${cleanedCount}个`);
    }
  }

  /**
   * 计算字符串哈希
   * @private
   */
  _hash(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return hash.toString(36);
  }

  /**
   * 获取缓存统计
   */
  getStats() {
    const hitRate = this.stats.totalRequests > 0
      ? ((this.stats.cacheHits / this.stats.totalRequests) * 100).toFixed(2)
      : '0.00';

    const semanticRate = this.stats.cacheHits > 0
      ? ((this.stats.semanticHits / this.stats.cacheHits) * 100).toFixed(2)
      : '0.00';

    return {
      ...this.stats,
      hitRate: `${hitRate}%`,
      semanticRate: `${semanticRate}%`,
      cacheSize: this.cache.size,
      maxSize: this.maxSize,
    };
  }

  /**
   * 清空缓存
   */
  clear() {
    this.cache.clear();
    this.accessOrder = [];
    logger.info('[SmartPlanCache] 缓存已清空');
  }

  /**
   * 销毁缓存
   */
  destroy() {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = null;
    }

    this.clear();
    logger.info('[SmartPlanCache] 缓存已销毁');
  }
}

module.exports = { SmartPlanCache, CacheEntry };

File: ios-app/ChainlessChain/Features/AI/Advanced/SmartPlanCache.swift
import Foundation
import CoreCommon

// MARK: - SmartPlanCache
/// Smart task plan cache with semantic similarity matching
/// Uses LRU eviction and cosine similarity for intelligent cache hits
///
/// Features:
/// 1. LLM Embedding vectorization
/// 2. Cosine similarity matching
/// 3. LRU eviction strategy
/// 4. Statistics and monitoring
///
/// Version: v1.7.0
/// Date: 2026-02-10

// MARK: - Cache Entry

/// LRU cache entry for task plans
struct PlanCacheEntry: Identifiable {
    let id: String
    let key: String           // Hash of original request (for quick lookup)
    let request: String       // Original request text
    let plan: TaskPlan        // Cached task plan
    let embedding: [Float]    // Request embedding vector
    var hits: Int             // Hit count
    let createdAt: Date
    var lastHitAt: Date?
    var lastAccessAt: Date

    init(key: String, request: String, plan: TaskPlan, embedding: [Float]) {
        self.id = UUID().uuidString.prefix(8).lowercased()
        self.key = key
        self.request = request
        self.plan = plan
        self.embedding = embedding
        self.hits = 0
        self.createdAt = Date()
        self.lastHitAt = nil
        self.lastAccessAt = Date()
    }
}

/// Task plan structure
struct TaskPlan: Codable {
    var description: String
    var subtasks: [TaskStep]
    var metadata: [String: String]?

    struct TaskStep: Codable {
        var tool: String
        var title: String
        var params: [String: AnyCodableValue]?
        var timeout: Int?
        var retries: Int?
        var retryDelay: Int?
        var regenerateParams: Bool?
        var regenerateCode: Bool?
        var strictSyntaxCheck: Bool?
    }
}

// MARK: - Cache Statistics

/// Statistics for smart plan cache
struct PlanCacheStats {
    var totalRequests: Int = 0
    var cacheHits: Int = 0
    var cacheMisses: Int = 0
    var semanticHits: Int = 0      // Semantic matching hits
    var exactHits: Int = 0         // Exact matching hits
    var evictions: Int = 0         // Eviction count
    var embeddingCalls: Int = 0    // Embedding API calls
    var embeddingFailures: Int = 0

    var hitRate: Double {
        guard totalRequests > 0 else { return 0 }
        return Double(cacheHits) / Double(totalRequests) * 100
    }

    var semanticRate: Double {
        guard cacheHits > 0 else { return 0 }
        return Double(semanticHits) / Double(cacheHits) * 100
    }
}

// MARK: - Smart Plan Cache

/// Smart plan cache with semantic similarity matching
@MainActor
class SmartPlanCache: ObservableObject {
    // MARK: - Configuration

    struct Config {
        var maxSize: Int = 1000                        // Maximum cache entries
        var similarityThreshold: Float = 0.85          // Similarity threshold (0-1)
        var ttl: TimeInterval = 7 * 24 * 60 * 60       // 7 days TTL
        var enabled: Bool = true
    }

    // MARK: - Properties

    private var config: Config
    private var cache: [String: PlanCacheEntry] = [:]  // key -> entry
    private var accessOrder: [String] = []             // LRU order
    private var cleanupTimer: Timer?
    private let logger = Logger.shared

    @Published private(set) var stats = PlanCacheStats()
    @Published private(set) var cacheSize: Int = 0

    // LLM Manager reference for embeddings
    private weak var llmManager: LLMManager?

    // Vocabulary for TF-IDF fallback
    private let vocabulary = [
        "create", "add", "build", "generate", "implement",
        "update", "modify", "change", "edit", "refactor",
        "delete", "remove", "clean", "clear",
        "test", "check", "validate", "verify",
        "file", "code", "function", "class", "component",
        "api", "database", "server", "client",
        "feature", "bug", "fix", "improve", "optimize"
    ]

    // MARK: - Singleton

    static let shared = SmartPlanCache()

    // MARK: - Initialization

    init(config: Config = Config()) {
        self.config = config

        // Start cleanup timer (every hour)
        startCleanupTimer()

        logger.info("[SmartPlanCache] Initialized with maxSize: \(config.maxSize), threshold: \(config.similarityThreshold)")
    }

    /// Set LLM manager for embeddings
    func setLLMManager(_ manager: LLMManager) {
        self.llmManager = manager
    }

    // MARK: - Public Methods

    /// Get cached plan (supports semantic matching)
    /// - Parameter request: User request
    /// - Returns: Cached plan or nil
    func get(_ request: String) async -> TaskPlan? {
        guard config.enabled else { return nil }

        stats.totalRequests += 1

        // 1. Try exact match (fast path)
        let exactKey = hash(request)
        if let entry = cache[exactKey], !isExpired(entry) {
            recordHit(key: exactKey, type: .exact)
            stats.exactHits += 1
            stats.cacheHits += 1

            logger.debug("[SmartPlanCache] Exact hit: \(entry.id), hits: \(entry.hits)")
            return entry.plan
        }

        // 2. Semantic similarity matching
        if let semanticPlan = await semanticSearch(request) {
            stats.semanticHits += 1
            stats.cacheHits += 1
            return semanticPlan
        }

        // 3. Cache miss
        stats.cacheMisses += 1
        logger.debug("[SmartPlanCache] Cache miss")

        return nil
    }

    /// Set cache entry
    /// - Parameters:
    ///   - request: User request
    ///   - plan: Task plan
    func set(_ request: String, plan: TaskPlan) async {
        guard config.enabled else { return }

        let key = hash(request)

        // Check if already exists
        if cache[key] != nil {
            logger.debug("[SmartPlanCache] Updating existing cache: \(key)")
            cache[key]?.plan = plan
            cache[key]?.lastAccessAt = Date()
            updateAccessOrder(key)
            return
        }

        // Calculate embedding
        guard let embedding = await getEmbedding(request) else {
            logger.warning("[SmartPlanCache] Embedding failed, skipping cache")
            return
        }

        // Create new entry
        let entry = PlanCacheEntry(key: key, request: request, plan: plan, embedding: embedding)

        // Check if cache is full
        if cache.count >= config.maxSize {
            evictLRU()
        }

        // Add to cache
        cache[key] = entry
        accessOrder.append(key)
        cacheSize = cache.count

        logger.debug("[SmartPlanCache] Cache added: \(entry.id), size: \(cache.count)/\(config.maxSize)")
    }

    /// Clear all cache
    func clear() {
        cache.removeAll()
        accessOrder.removeAll()
        cacheSize = 0
        logger.info("[SmartPlanCache] Cache cleared")
    }

    /// Get cache statistics
    func getStats() -> [String: Any] {
        return [
            "totalRequests": stats.totalRequests,
            "cacheHits": stats.cacheHits,
            "cacheMisses": stats.cacheMisses,
            "exactHits": stats.exactHits,
            "semanticHits": stats.semanticHits,
            "evictions": stats.evictions,
            "embeddingCalls": stats.embeddingCalls,
            "embeddingFailures": stats.embeddingFailures,
            "hitRate": String(format: "%.2f%%", stats.hitRate),
            "semanticRate": String(format: "%.2f%%", stats.semanticRate),
            "cacheSize": cache.count,
            "maxSize": config.maxSize
        ]
    }

    /// Update configuration
    func updateConfig(_ newConfig: Config) {
        config = newConfig
        logger.info("[SmartPlanCache] Config updated: maxSize=\(config.maxSize), threshold=\(config.similarityThreshold)")
    }

    // MARK: - Private Methods

    /// Semantic search for similar plans
    private func semanticSearch(_ request: String) async -> TaskPlan? {
        guard let requestEmbedding = await getEmbedding(request) else {
            return nil
        }

        var bestMatch: PlanCacheEntry?
        var bestSimilarity: Float = 0

        // Traverse all cache entries, calculate similarity
        for entry in cache.values {
            if isExpired(entry) { continue }

            let similarity = cosineSimilarity(requestEmbedding, entry.embedding)

            if similarity > bestSimilarity && similarity >= config.similarityThreshold {
                bestSimilarity = similarity
                bestMatch = entry
            }
        }

        if let match = bestMatch {
            recordHit(key: match.key, type: .semantic)
            logger.debug("[SmartPlanCache] Semantic hit: \(match.id), similarity: \(String(format: "%.2f%%", bestSimilarity * 100)), hits: \(match.hits)")
            return match.plan
        }

        return nil
    }

    /// Get embedding vector for text
    private func getEmbedding(_ text: String) async -> [Float]? {
        stats.embeddingCalls += 1

        // Try LLM embedding first
        if let manager = llmManager {
            do {
                let embedding = try await manager.generateEmbedding(text)
                return embedding
            } catch {
                logger.debug("[SmartPlanCache] LLM embedding failed: \(error.localizedDescription)")
            }
        }

        // Fallback to TF-IDF vector
        logger.debug("[SmartPlanCache] Using TF-IDF fallback")
        return simpleTFIDFVector(text)
    }

    /// Simple TF-IDF vectorization (fallback)
    private func simpleTFIDFVector(_ text: String) -> [Float] {
        // Tokenize
        let tokens = text.lowercased()
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { $0.count > 2 }

        // Calculate term frequency
        var freq: [String: Int] = [:]
        for token in tokens {
            freq[token, default: 0] += 1
        }

        // Build vector based on vocabulary
        let totalTokens = Float(tokens.count)
        var vector: [Float] = []

        for word in vocabulary {
            let tf = totalTokens > 0 ? Float(freq[word] ?? 0) / totalTokens : 0
            vector.append(tf)
        }

        return vector
    }

    /// Calculate cosine similarity
    private func cosineSimilarity(_ vec1: [Float], _ vec2: [Float]) -> Float {
        guard vec1.count == vec2.count, !vec1.isEmpty else { return 0 }

        var dotProduct: Float = 0
        var norm1: Float = 0
        var norm2: Float = 0

        for i in 0..<vec1.count {
            dotProduct += vec1[i] * vec2[i]
            norm1 += vec1[i] * vec1[i]
            norm2 += vec2[i] * vec2[i]
        }

        let denominator = sqrt(norm1) * sqrt(norm2)

        guard denominator > 0 else { return 0 }

        return dotProduct / denominator
    }

    /// Record cache hit
    private enum HitType {
        case exact, semantic
    }

    private func recordHit(key: String, type: HitType) {
        cache[key]?.hits += 1
        cache[key]?.lastHitAt = Date()
        cache[key]?.lastAccessAt = Date()
        updateAccessOrder(key)
    }

    /// Update access order (LRU)
    private func updateAccessOrder(_ key: String) {
        if let index = accessOrder.firstIndex(of: key) {
            accessOrder.remove(at: index)
        }
        accessOrder.append(key)
    }

    /// LRU eviction
    private func evictLRU() {
        guard !accessOrder.isEmpty else { return }

        let lruKey = accessOrder.removeFirst()
        if let evicted = cache.removeValue(forKey: lruKey) {
            stats.evictions += 1
            cacheSize = cache.count
            logger.debug("[SmartPlanCache] LRU evicted: \(evicted.id), hits: \(evicted.hits)")
        }
    }

    /// Check if entry is expired
    private func isExpired(_ entry: PlanCacheEntry) -> Bool {
        return Date().timeIntervalSince(entry.createdAt) > config.ttl
    }

    /// Cleanup expired entries
    private func cleanupExpired() {
        var cleanedCount = 0

        for (key, entry) in cache {
            if isExpired(entry) {
                cache.removeValue(forKey: key)
                if let index = accessOrder.firstIndex(of: key) {
                    accessOrder.remove(at: index)
                }
                cleanedCount += 1
            }
        }

        if cleanedCount > 0 {
            cacheSize = cache.count
            logger.info("[SmartPlanCache] Cleaned expired entries: \(cleanedCount)")
        }
    }

    /// Calculate string hash
    private func hash(_ str: String) -> String {
        var hash: Int = 0
        for char in str.unicodeScalars {
            hash = ((hash << 5) &- hash) &+ Int(char.value)
        }
        return String(hash, radix: 36)
    }

    /// Start cleanup timer
    private func startCleanupTimer() {
        cleanupTimer = Timer.scheduledTimer(withTimeInterval: 60 * 60, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.cleanupExpired()
            }
        }
    }

    deinit {
        cleanupTimer?.invalidate()
    }
}

// MARK: - Extension for mutable plan property

extension PlanCacheEntry {
    var plan: TaskPlan {
        get { _plan }
        set { _plan = newValue }
    }
    private var _plan: TaskPlan {
        get { _planStorage }
        set { _planStorage = newValue }
    }
}

// Using a separate storage since struct is immutable by default
private var _planStorage: [String: TaskPlan] = [:]

extension PlanCacheEntry {
    fileprivate var _planStorage: TaskPlan {
        get { _planStorage[id] ?? plan }
        set { _planStorage[id] = newValue }
    }
}
